hamilton/ad_hoc_utils.py

```diff
@@ -57,3 +58,20 @@ def create_temporary_module(*functions: Callable, module_name: str = None) -> Mo
         setattr(module, fn_name, fn)
     sys.modules[module_name] = module
     return module
+
+
+def module_from_source(source: str) -> ModuleType:
```

Comment:
Without this function, the `CacheHook` doesn't work in notebooks because it won't be able to use `inspect` to get the source code and create a hash.

The alternative would be to use the functions's `__code__` attribute, but this would dependent on the Python version. It might not be an issue because the cache's pickling is already Python version dependent

Reply:
Where is this used? Think I'm missing it... But that makes sense I think. You're talking about temporary modules, right?

With temporary modules you'd have to special-case them, go through every function and use `inspect` on them.